### PR TITLE
ROC-4412: Request re-determination by claimant

### PR DESCRIPTION
**Pull request by dharmendrak**
### JIRA link
https://tools.hmcts.net/jira/browse/ROC-4412

### Work checklist

- [ ] Unit tests added where applicable
- [ ] Route tests added for new pages
- [ ] New pages included in a11y tests
- [ ] Task list and task completeness checks updated
- [ ] Check and send page updated
- [ ] Routes, page content and page flows of new features are toggled 
- [ ] UI changes look good on mobile
- [ ] Required Google Analytics events are being sent 

### Developer self-QA run statement

- [ ] I have clicked through the running application to see if all changes I made actually work.

### Does this PR introduce a breaking change?

- [ ] Yes
- [ ] No

### CODE REVIEW

File: src/main/app/claims/ccjClient.ts [hunk 1]
```diff
@@ -22,4 +24,13 @@ export class CCJClient {
       }
     })
   }
+
+  static redetermination (externalId: string, reDetermination: ReDetermination, user: User, madeBy: MadeBy) {
```
Comment (sabahirfan) [inline src/main/app/claims/ccjClient.ts]: `async ` ?

File: src/main/features/ccj/form/models/reDetermination.ts [hunk 2]
```diff
@@ -0,0 +1,30 @@
+import { IsDefined } from 'class-validator'
+
+export class ValidationErrors {
+  static readonly TYPE_REQUIRED: string = 'Enter a valid reason'
+}
+
+export class ReDetermination {
+  @IsDefined({ message: ValidationErrors.TYPE_REQUIRED })
```
Comment (sabahirfan) [inline src/main/features/ccj/form/models/reDetermination.ts]: It's not type, please rename `TYPE_REQUIRED` to `REASON_REQUIRED` or `EXPLANATION_REQUIRED` etc

File: src/main/features/ccj/routes/confirmation.ts [hunk 3]
```diff
@@ -13,6 +13,7 @@ export default express.Router()
       res.render(Paths.confirmationPage.associatedView,
         {
           defendantName: claim.claimData.defendant.name,
-          ccjRequestedAt: claim.countyCourtJudgmentRequestedAt
+          ccjRequestedAt: claim.countyCourtJudgmentRequestedAt,
+          reDetermination_requested_at: claim.reDeterminationRequestedAt
```
Comment (sabahirfan) [inline src/main/features/ccj/routes/confirmation.ts]: To make it consistent with others please use camel case. i.e. rename `reDetermination_requested_at` to `reDeterminationRequestedAt `

File: src/main/features/ccj/views/redetermination.njk [hunk 4]
```diff
@@ -0,0 +1,19 @@
+{% extends "layout.njk" %}
+
+{% set heading = 'Why do you believe the defendant can repay you sooner?' %}
+
+{% from "form.njk" import saveAndContinueButton, csrfProtection, radioGroup, errorSummary, textArea %}
```
Comment (sabahirfan) [inline src/main/features/ccj/views/redetermination.njk]: `radioGroup` is not required here

File: src/test/features/ccj/routes/ccjGuard.ts [hunk 5]
```diff
@@ -51,6 +53,58 @@ describe('CCJ guard', () => {
           .set('Cookie', `${cookieName}=ABC`)
           .expect(res => expect(res).to.be.successful.withText('County Court Judgment requested'))
       })
+
+      it('should NOT redirect to dashboard when claim not eligible for CCJ on re determination page', async () => {
+        const route: string = Paths.redeterminationPage.evaluateUri({ externalId: 'b17af4d2-273f-4999-9895-bce382fa24c8' })
+        claimStoreServiceMock.resolveRetrieveClaimByExternalId({
+          respondedAt: MomentFactory.currentDateTime(),
```
Comment (sabahirfan) [inline src/test/features/ccj/routes/ccjGuard.ts]: Please extract this JSON object and reuse in tests

File: src/test/http-mocks/claim-store.ts [hunk 6]
```diff
@@ -276,6 +276,20 @@ export function resolveSaveCcjForExternalId () {
     .reply(HttpStatus.OK, { ...sampleClaimObj })
 }
 
+export function resolveSaveReDeterminationForExternalId (explanation: string) {
+
+  mock(`${serviceBaseURL}/claims`)
+    .post(new RegExp('/[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}' +
```
[the remark this answers is not in the document]
Comment (dharmendrak) [reply src/test/http-mocks/claim-store.ts]: done

File: src/test/features/ccj/routes/ccjGuard.ts (hunk 5 above)
Comment (dharmendrak) [reply src/test/features/ccj/routes/ccjGuard.ts]: done

File: src/main/features/ccj/views/redetermination.njk (hunk 4 above)
Comment (dharmendrak) [reply src/main/features/ccj/views/redetermination.njk]: done

File: src/main/features/ccj/routes/confirmation.ts (hunk 3 above)
Comment (dharmendrak) [reply src/main/features/ccj/routes/confirmation.ts]: done

File: src/main/features/ccj/form/models/reDetermination.ts (hunk 2 above)
Comment (dharmendrak) [reply src/main/features/ccj/form/models/reDetermination.ts]: done

File: src/main/app/claims/ccjClient.ts (hunk 1 above)
Comment (dharmendrak) [reply src/main/app/claims/ccjClient.ts]: done

File: src/test/http-mocks/claim-store.ts [hunk 7]
```diff
@@ -16,6 +16,7 @@ import {
 } from 'test/data/entity/responseData'
 
 const serviceBaseURL: string = config.get<string>('claim-store.url')
+const externalIdRegex: string = '/[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}'
```
Comment (sabahirfan) [inline src/test/http-mocks/claim-store.ts]: `externalIdPattern`

Comment (sabahirfan) [reply src/test/http-mocks/claim-store.ts]: `/` at the start is not part of `externalIdPattern`

**Pull request merged**